File: ch01/sec1.7/numbers/symb.py
from functools import reduce

def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return abs(a)

class Fraction:
    def __init__(self, numerator, denominator):
        if denominator == 0:
            raise ValueError("Denominator cannot be zero.")
        common = gcd(numerator, denominator)
        self.numerator = numerator // common
        self.denominator = denominator // common

    def __str__(self):
        return f"{self.numerator}/{self.denominator}"

    def __add__(self, other):
        num = self.numerator * other.denominator + self.denominator * other.numerator
        denom = self.denominator * other.denominator
        return Fraction(num, denom)

    def __sub__(self, other):
        num = self.numerator * other.denominator - self.denominator * other.numerator
        denom = self.denominator * other.denominator
        return Fraction(num, denom)

    def __mul__(self, other):
        num = self.numerator * other.numerator
        denom = self.denominator * other.denominator
        return Fraction(num, denom)

    def __truediv__(self, other):
        num = self.numerator * other.denominator
        denom = self.denominator * other.numerator
        return Fraction(num, denom)

class SymbolicFraction(Fraction):
    def __init__(self, numerator, denominator):
        super().__init__(numerator, denominator)
        self.symbolic_expr = f"{self.numerator}/{self.denominator}"

    def __add__(self, other):
        expr = f"({self.symbolic_expr} + {other.symbolic_expr})"
        num = self.numerator * other.denominator + self.denominator * other.numerator
        denom = self.denominator * other.denominator
        result = SymbolicFraction(num, denom)
        result.symbolic_expr = expr
        return result

    def __sub__(self, other):
        expr = f"({self.symbolic_expr} - {other.symbolic_expr})"
        num = self.numerator * other.denominator - self.denominator * other.numerator
        denom = self.denominator * other.denominator
        result = SymbolicFraction(num, denom)
        result.symbolic_expr = expr
        return result

    def __mul__(self, other):
        expr = f"({self.symbolic_expr} * {other.symbolic_expr})"
        num = self.numerator * other.numerator
        denom = self.denominator * other.denominator
        result = SymbolicFraction(num, denom)
        result.symbolic_expr = expr
        return result

    def __truediv__(self, other):
        expr = f"({self.symbolic_expr} / {other.symbolic_expr})"
        num = self.numerator * other.denominator
        denom = self.denominator * other.numerator
        result = SymbolicFraction(num, denom)
        result.symbolic_expr = expr
        return result

    def __str__(self):
        return f"{self.symbolic_expr} = {self.numerator}/{self.denominator}"

    # comparison operations (returns boolean)
    def is_equal_to(self, other):
        return self.numerator * other.denominator == self.denominator * other.numerator

    def is_greater_than(self, other):
        return self.numerator * other.denominator > self.denominator * other.numerator

    def is_less_than(self, other):
        return self.numerator * other.denominator < self.denominator * other.numerator

    # conditional expression
    def if_greater_than(self, other, result_if_true, result_if_false):
        if self.is_greater_than(other):
            return result_if_true
        else:
            return result_if_false

# example
f1 = SymbolicFraction(1, 2)
f2 = SymbolicFraction(3, 4)

# symbolic representation without evaluation
symbolic_sum = f1 + f2
print(symbolic_sum)  # output: "(1/2 + 3/4) = 5/4"

# conditional check
result = f1.if_greater_than(f2, "f1 is greater", "f2 is greater")
print(result)  # output: "f2 is greater"

File: ch01/sec1.7/numbers/symb.c
#include <stdio.h>
#include <string.h>

typedef struct {
    int numerator;
    int denominator;
    char symbolic_expr[50];
} SymbolicFraction;

int gcd(int a, int b) {
    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

SymbolicFraction simplify(SymbolicFraction frac) {
    int common = gcd(frac.numerator, frac.denominator);
    frac.numerator /= common;
    frac.denominator /= common;
    return frac;
}

SymbolicFraction add(SymbolicFraction f1, SymbolicFraction f2) {
    SymbolicFraction result;
    result.numerator = f1.numerator * f2.denominator + f1.denominator * f2.numerator;
    result.denominator = f1.denominator * f2.denominator;
    result = simplify(result);
    snprintf(result.symbolic_expr, sizeof(result.symbolic_expr), "(%s + %s)", f1.symbolic_expr, f2.symbolic_expr);
    return result;
}

SymbolicFraction subtract(SymbolicFraction f1, SymbolicFraction f2) {
    SymbolicFraction result;
    result.numerator = f1.numerator * f2.denominator - f1.denominator * f2.numerator;
    result.denominator = f1.denominator * f2.denominator;
    result = simplify(result);
    snprintf(result.symbolic_expr, sizeof(result.symbolic_expr), "(%s - %s)", f1.symbolic_expr, f2.symbolic_expr);
    return result;
}

SymbolicFraction multiply(SymbolicFraction f1, SymbolicFraction f2) {
    SymbolicFraction result;
    result.numerator = f1.numerator * f2.numerator;
    result.denominator = f1.denominator * f2.denominator;
    result = simplify(result);
    snprintf(result.symbolic_expr, sizeof(result.symbolic_expr), "(%s * %s)", f1.symbolic_expr, f2.symbolic_expr);
    return result;
}

SymbolicFraction divide(SymbolicFraction f1, SymbolicFraction f2) {
    SymbolicFraction result;
    result.numerator = f1.numerator * f2.denominator;
    result.denominator = f1.denominator * f2.numerator;
    result = simplify(result);
    snprintf(result.symbolic_expr, sizeof(result.symbolic_expr), "(%s / %s)", f1.symbolic_expr, f2.symbolic_expr);
    return result;
}

void print_fraction(SymbolicFraction f) {
    printf("%s = %d/%d\n", f.symbolic_expr, f.numerator, f.denominator);
}

// comparison
int is_equal(SymbolicFraction f1, SymbolicFraction f2) {
    return f1.numerator * f2.denominator == f1.denominator * f2.numerator;
}

int is_greater(SymbolicFraction f1, SymbolicFraction f2) {
    return f1.numerator * f2.denominator > f1.denominator * f2.numerator;
}

int is_less(SymbolicFraction f1, SymbolicFraction f2) {
    return f1.numerator * f2.denominator < f1.denominator * f2.numerator;
}

// conditionals
void if_greater(SymbolicFraction f1, SymbolicFraction f2, char* result_if_true, char* result_if_false) {
    if (is_greater(f1, f2)) {
        printf("%s\n", result_if_true);
    } else {
        printf("%s\n", result_if_false);
    }
}

int main() {
    SymbolicFraction f1 = {1, 2, "1/2"};
    SymbolicFraction f2 = {3, 4, "3/4"};

    // addition
    SymbolicFraction result = add(f1, f2);
    print_fraction(result); // output: "(1/2 + 3/4) = 5/4"

    // conditional check
    if_greater(f1, f2, "f1 is greater", "f2 is greater"); // output: "f2 is greater"

    return 0;
}
